### PR TITLE
Add initial version of Beman software standard.

### PR DESCRIPTION
**Pull request by camio**

### CODE REVIEW

File: docs/beman-standard.md
Comment (JeffGarland) [inline docs/beman-standard.md:74]: Should we list other options we've already discussed aka MIT?

**Review by JeffGarland — APPROVED** [review]

**Review by bretbrownjr — COMMENTED** [review]
Did I miss a discussion about why the boost license is recommended over apache and mit?

I'm not necessarily opposed, but that wasn't the thinking a couple weeks ago. Also it seems to go against the "use the normal tech" standard given it's relatively less popular than the other two.

Regardless, if that detail is merged, we'll need followup issues or even PRs to update the example repo. I expect we should be iterating on it until it's stable, and then turning it into a project template.

**Comment by camio** [comment]
> Did I miss a discussion about why the boost license is recommended over apache and mit?

No, not at all. I put the Boost license as a recommendation more because we need a recommendation than consideration of which license is preferred. Do you have an opinion on this Bret? If so, it may be good to create a discourse thread where the options can be weighed. I'm more than happy to update this PR to reflect that.

Incidentally, it looks like `beman` has Apache2+LLVM, `example` has MIT, and this PR mentions Boost. :)

To launch the discussion, perhaps [LICENSE.CRITERIA], and [CORE.INDUSTRY_STANDARD] can help guide a decision.

**Comment by bretbrownjr** [comment]
I don't have a lot of opinions about licensing, personally. I would expect to see some justification for using Boost over the more popular licenses, but I wouldn't block this PR on that personal opinion. It can always be clarified (or ignored) later.

**Comment by camio** [comment]
I changed the recommended license to Apache-2.0 WITH LLVM-exception.  I also created [this topic](https://discourse.boost.org/t/beman-standard-drafted/118) which has the rationale and invites feedback on the draft as a whole.

File: docs/beman-standard.md
Comment (rmpowell77) [inline docs/beman-standard.md:81]: Do we need to have anything here about liability?

File: docs/beman-standard.md
Comment (JeffGarland) [inline docs/beman-standard.md:85]: I think this has to be a recommendation instead of a requirement -- or we have to have some way to reference things from Beman without impacting the core library.  For example the scnlib that we're currently including already has a name that doesn't follow the convention.

File: docs/beman-standard.md
Comment (rmpowell77) [inline docs/beman-standard.md:125]: Maybe we could add something about Paper number?

```
**[README.PAPER]** RECOMMENDATION: The `README.md` should contain the reference to
the paper number the library is covering.

# Paper: p2300r4
```

**Review by rmpowell77 — APPROVED** [review]

File: docs/beman-standard.md
Comment (JeffGarland) [inline docs/beman-standard.md:169]: We probably don't want to use a famous actor with a non-existent email address...

File: docs/beman-standard.md
Comment (JeffGarland) [inline docs/beman-standard.md:134]: it's a fairly usual convention in a header-only library to put the impl headers in a <shortname>/impl/ 

If we are going to put headers under src then we should use the pitchfork (from vector-of-bool) recommendations for what's called separated header placement:



```

In separated placement, there are two source directories, [§ 2.2 include/](https://api.csswg.org/bikeshed/?force=1&url=https://raw.githubusercontent.com/vector-of-bool/pitchfork/develop/data/spec.bs#tld.include) and [§ 2.3 src/](https://api.csswg.org/bikeshed/?force=1&url=https://raw.githubusercontent.com/vector-of-bool/pitchfork/develop/data/spec.bs#tld.src). The include/ directory is designated to contain the public headers of the library, while the src/ directory is designated to contain the compilable source code and private headers.In separated placement, there are two source directories, [§ 2.2 include/](https://api.csswg.org/bikeshed/?force=1&url=https://raw.githubusercontent.com/vector-of-bool/pitchfork/develop/data/spec.bs#tld.include) and [§ 2.3 src/](https://api.csswg.org/bikeshed/?force=1&url=https://raw.githubusercontent.com/vector-of-bool/pitchfork/develop/data/spec.bs#tld.src). The include/ directory is designated to contain the public headers of the library, while the src/ directory is designated to contain the compilable source code and private headers.

```

File: docs/beman-standard.md
Comment (camio) [reply docs/beman-standard.md:81]: Do you have any wording in mind?

File: docs/beman-standard.md
Comment (camio) [reply docs/beman-standard.md:85]: @eliaskosunen would you object to calling your library Beman.Scnlib or something like that if it gets incorporated into the Beman project.

File: docs/beman-standard.md
Comment (camio) [reply docs/beman-standard.md:125]: Added [here](https://github.com/beman-project/beman/pull/18/commits/0f86bf9ca3bfbfd3acb4799e355bc6cb728dcc9f), but with a bit different wording:

- Allow for multiple papers to be implemented by a single repo
- Link to the official version of the paper
- Use the paper's title and put the number in parenthesis since the latter is committee speak.

File: docs/beman-standard.md
Comment (camio) [reply docs/beman-standard.md]: Fixed in https://github.com/beman-project/beman/pull/18/commits/d858d8eadc193713d03d3670559e260799470c37

File: docs/beman-standard.md
Comment (bretbrownjr) [reply docs/beman-standard.md:85]: It's a good practice to have names all correspond across the entire library. In the sense of physical design and logical design "matching".

Including: the repo name, project name, header include targets, library target name in CMake, library file name, C++ namespace name, macro name prefix if applicable, extern C names, and maybe some other things.

Do we want `beman::foobar::some_range` and `find_package(beman-foobar)` and `#include <beman/foobar/core.hpp>` and `libbeman_foobar.a` and so on?

What if I library ends up gaining its own life beyond the standard? Some libraries could reasonably be declined for standardization and "only" end up being very popular libraries like nlohmanm-json, googletest, or cli11. Do we expect a "hard fork" in those cases, renaming everything to de-bemanize it?

I don't see a technical advantage to the Beman to level scope compared to just using a unique enough library name. I would expect good docs and developer support and a helpful community would speak for itself if branding is the concern. But I guess I'm an engineer and not a marketing guy.

File: docs/beman-standard.md
Comment (neatudarius) [inline docs/beman-standard.md:127]: Should we also state to not include actual paper code (eg markdown, html, pdf etc)? I think each author can have a separate repo for that, but only include relevant implementation in Beman.

File: docs/beman-standard.md
Comment (neatudarius) [reply docs/beman-standard.md:143]: I prefer the `libc++` convention.

**Review by neatudarius — APPROVED** [review]
LGTM 👍

File: docs/beman-standard.md
Comment (bretbrownjr) [reply docs/beman-standard.md:143]: I'm mostly flexible. Fairly easy-to-meet technical asks:

* Includes in-repo must be identical to cross-project includes. Let's always include `foobar/core.hpp` and not sometimes include `../foobar/core.hpp` even though it's the same header.
* Let's prefix *every* include target with the library name. Yes to `foobar/core.hpp`. No thanks to `bits/core.hpp`. No thanks to `detail/foobar/core.hpp`. `libstdc++` gets this wrong for whatever reason. Effectively if project is named `libbits`, it sort of overlaps with `libstdc++` in its include target namespace. If `stdc++` put private headers in `libstdc++/internal/string.h` or something like that, it would be unambiguous what's going on.
* It should be hard to accidentally ship a header you didn't mean to. Probably this means shipped and unshipped headers live in entirely different subdirectories somehow.

Note that tools like include-what-you-use get confused regardless of the not-for-public-use convention, especially because it is a convention and not a standard. Eventually C++ needs an official way to declare what headers provide what symbols. Among the many reasons I'm also working on [CPS](https://github.com/cps-org/cps) is that it makes sense to eventually declare some of that sort of thing in a declarative JSON interface instead of by making up conventions per project.

Anyway, I'll go with consensus on this detail. I tend to put all of the source in one directory, personally. Careful naming of include targets covers all the ambiguity, but @camio is right that you can accidentally include something you didn't intend to ship that way. I personally find that the tooling to prevent that sort of bug is really lacking, but good enough test coverage of already packaged libraries (see my conan PR for example tests) would protect against any accidents there.

File: docs/beman-standard.md
Comment (bretbrownjr) [reply docs/beman-standard.md:127]: If I were maintaining a library paper, I'd rather have everything in one place, personally.

Not that I have capacity for library papers these days, but if I were, I'd be tempted to just do an in-repo markdown file for the paper and convert to pdf when mailing. It might render less elegantlantly than LaTeX, but maybe I don't particularly care. I'd want a normal open source workflow for taking suggestions, and I'd generally be open to additional authors, though.

Comment (neatudarius) [reply docs/beman-standard.md:127]: I expect that authors have repos with papers and possibly experiments, at least for ideas that exist at this moment.

Also, I not find OK allowing others changes paper content. The paper is strictly a proposal from a group of authors (which can be extended if they want; otherwise it's a conservative group). I see Beman used only for implementing ideas.

TLDR: My opinion is that keeping papers inside Beman can diverge from authors purpose and from a direction already approved.

**Comment by pdimov** [comment]
> Did I miss a discussion about why the boost license is recommended over apache and mit?

This is why: https://pdimov.github.io/blog/2020/09/06/why-use-the-boost-license/

File: docs/beman-standard.md
Comment (camio) [reply docs/beman-standard.md:127]: Instances where an author produces both a paper and its Beman implementation will be common. In this case, a single repository allows for paper/implementation updates to be treated as a single unit of change which is great for reviews.  We can also provide a convenient paper template.

On the flip side, there are cases where one Beman library implements several papers (e.g. Beman.Optional26) or where the implementor is not the paper author.  We need to support those situations as well.

I think what's being proposed in this PR is flexible enough to handle these situations. Perhaps a future PR could provide recommendations where papers sources that cohabitate with a Beman repository should go.

File: docs/beman-standard.md
Comment (camio) [reply docs/beman-standard.md]: I changed this requirement to a recommendation. Some libraries have existing branding or have use cases outside the context of the Beman project. Such libraries should be permitted in the Beman project.

However, to address @bretbrownjr 's concern, we are keeping the requirement that Beman libraries have a unified interface (e.g. `#include <Beman/ShortName/foo.hpp`) so Beman users know exactly what is being proposed for the standard. Additionally, after this initial PR, I think we should have others that recommend unified CMake package names, library names, etc.

File: docs/beman-standard.md
Comment (camio) [reply docs/beman-standard.md:148]: I added [DIRECTORY.IMPLEMENTATION_HEADERS] which attempts to reflect the resolution we had at our last weekly meeting.

File: docs/beman-standard.md
Comment (rmpowell77) [reply docs/beman-standard.md:81]: Actually, I think we do not need a policy here.

**Review by rmpowell77 — APPROVED** [review]
I think this is a good direction to go.

**Review by neatudarius — APPROVED** [review]

File: docs/beman-standard.md
Comment (dabrahams) [inline docs/beman-standard.md:68]: I don't know what these bracketed things are supposed to mean, but I note that the name of this one seems oddly uncorrelated to the recommendation.

File: docs/beman-standard.md
Comment (dabrahams) [inline docs/beman-standard.md:72]: This is a rationale for your existing choices and may not belong in the main body of this document.  Otherwise it's redundant with the list of approved licenses.

File: docs/beman-standard.md
Comment (camio) [reply docs/beman-standard.md:68]: Those are entry identifiers used for referencing. Good catch on this one being poorly named! Fixed in https://github.com/beman-project/beman/commit/3a16df3c9614a22935e78964684123c5faac5a06.

File: docs/beman-standard.md
Comment (camio) [reply docs/beman-standard.md:72]: I agree this entry is awkward. The intent is to provide guidance to those wishing to get another license approved.

I see two ways to improve the situation (are there others?):

1. Maintain a separate Beman standard rationale document. Users of the standard benefit since rationale is not usually pertinent. The drawback is that the rationale document will either be sparse (we rarely *need* to agree on rationale) or will cost us time debating rationale wording.
2. Put the rationale inline. It may help with buy-in and facilitate improvements.

File: docs/beman-standard.md
Comment (JeffGarland) [reply docs/beman-standard.md:127]: > TLDR: My opinion is that keeping papers inside Beman can diverge from authors purpose and from a direction already approved.

I think it's more likely that the impl will diverge if it's kept in a different place.  We see this all the time where the paper changes something -- like the name of interface due to committee bikeshed -- but the implementation isn't updated.  I'd like to see both in the same place to start.

Obviously I'm commenting on a closed PR, but i do expect follow on changes as we learn.

Comment (JeffGarland) [reply docs/beman-standard.md:127]: > If I were maintaining a library paper, I'd rather have everything in one place, personally.
> 
> Not that I have capacity for library papers these days, but if I were, I'd be tempted to just do an in-repo markdown file for the paper and convert to pdf when mailing. It might render less elegantlantly than LaTeX, but maybe I don't particularly care. I'd want a normal open source workflow for taking suggestions, and I'd generally be open to additional authors, though.

This actually brings up tooling support. M. Park markdown is what many use -- we should look into if he uses 'make targets' or how that's done.  Probably someone here knows.

I use .org mode, and we've been discussing with Steve D more direct support for exporting M. Park markdown.  As it is, I exported to html for my last paper.  One really nice thing about .org is the paper renders in repo nicely: https://github.com/JeffGarland/wg21work/blob/main/monadic_expected/monadic_functions_for_expected.org

**Comment by JeffGarland** [comment]
> > Did I miss a discussion about why the boost license is recommended over apache and mit?
> 
> This is why: https://pdimov.github.io/blog/2020/09/06/why-use-the-boost-license/

Thanks this is a good reminder of why the Boost license is so good. Unfortunately @ldionne mentioned that the LLVM lawyers have some issues with it.  What @ldionne would like to be able to do is take say the unit tests from the proposed library, munge them a bit and turn them into LLVM reference tests.  This has the impact that all 3 major implementations immediately get access to these tests because they all run LLVM tests. This feels like a lawyer problem, not a legal problem -- but here we are.

File: docs/beman-standard.md
Comment (dabrahams) [reply docs/beman-standard.md]: You might consider requiring dual-licensing.  IIUC the Boost License is less constraining than the other one, with many fewer [redistribution requirements (section 4)](https://apache.org/licenses/LICENSE-2.0).   FWIW, Beman really cared about lowering the bar for usage.

File: docs/beman-standard.md
Comment (dabrahams) [reply docs/beman-standard.md:127]: M. Park Markdown?
[This](https://mpark.github.io/programming/2018/11/16/how-i-format-my-cpp-papers/) seems to indicate M. Park is using whatever pandoc supports

Comment (JeffGarland) [reply docs/beman-standard.md:127]: yeah it looks like he's using pandoc -- which supports a bunch of markdown flavors and probably allows some sort of configuration of the conversion of his special elements.

**Comment by pdimov** [comment]
Apache w/ LLVM exception is also fine, but MIT is not. I link [this post](https://www.reddit.com/r/cpp/comments/gr18ig/comment/frxbit4/) by Stephan T. Lavavej where he says the same thing (speaking as a stdlib maintainer.)

If the reference implementations in this project are intended to be usable as-is in standard libraries, without clean room rewriting, they have to use a license that permits it. In addition to Boost and Apache w/ LLVM exceptions, I only know of one other relatively well-known such license - the ZLib license (but you'll need to run that by a lawyer because I'm not aware of it having been evaluated from the perspective of using code covered by it in a stdlib implementation.)

**Comment by JeffGarland** [comment]
> Apache w/ LLVM exception is also fine, but MIT is not. 

Thx - good to know the Microsoft position. I'm starting to think we should exclude MIT since it's not liberal enough. @inbal2l what are your thoughts?  Still hoping @ldionne will weigh in on why LLVM doesn't like Boost.

**Comment by ldionne** [comment]
> > > Did I miss a discussion about why the boost license is recommended over apache and mit?
> > 
> > 
> > This is why: https://pdimov.github.io/blog/2020/09/06/why-use-the-boost-license/
> 
> Thanks this is a good reminder of why the Boost license is so good. Unfortunately @ldionne mentioned that the LLVM lawyers have some issues with it. What @ldionne would like to be able to do is take say the unit tests from the proposed library, munge them a bit and turn them into LLVM reference tests. This has the impact that all 3 major implementations immediately get access to these tests because they all run LLVM tests. This feels like a lawyer problem, not a legal problem -- but here we are.

Sorry, I'm late to the party here but I feel like it's important for me to clarify what I said.

LLVM uses the Apache-with-LLVM-exception license. The MSVC STL also uses that license (I think they are dual licensed). I mentioned to @JeffGarland and others that because of this, the path of least resistance for libc++ and MSVC STL to reuse code from the Beman Project would be to use that license, which would certainly sidestep some license-related considerations.

However, it is not the only license that would allow us to use code from the Beman Project -- it's just the one that makes it the easiest. In particular, I don't think LLVM or their lawyers have anything against the Boost license (or any other license) specifically. It's simply that using anything but the explicitly approved Apache-with-LLVM-exception license requires a bit of scrutiny, whereas using code under the Apache-with-LLVM-exception license is very straightforward.

**Pull request merged**
